### ISSUE TITLE
PROT_EXEC flag is mandatory for allocating memeory

### ISSUE THREAD

**Issue by FalconBsp** [RadeonOpenCompute/ROCT-Thunk-Interface#8]
allocate_exec_aligned_memory needs PROT_EXEC needs to be enabled. 

/dev/shm mounts with "noexec" option because of security reason. On the other hand, it seems that ROCm uses POSIX shared memory and mmap() it with "PROT_EXEC" option enabled, and mmap() failed with EPERM. Is it really needed to execute mmap() for POSIX shared memory with "PROT_EXEC" option?

Why we need PROT_EXEC needs for allocation of memory?

**Comment by fxkamd** [comment]
On APUs, CPU and GPU use the same page tables (through IOMMUv2). This means, for the GPU to be able to execute instructions from memory, the execute permission must be granted in the CPU page table. I believe this is needed for memory used for command queues and shader instructions.

The function allocate_exec_aligned_memory uses posix_memalign to allocate memory and memprotect to change the permissions. As far as I know, this should use anonymous memory, not memory mapped from /dev/shm.

Where are you seeing mmap of /dev/shm?

**Comment by SunilUttarwar** [comment]
We are observing in issue with clinfo or any OpenCL sample program with strict security setting environment.

**Comment by tstellar** [comment]
> We are observing in issue with clinfo or any OpenCL sample program with strict security setting environment.

Can you elaborate more on this?  Does this help: https://github.com/RadeonOpenCompute/ROCT-Thunk-Interface/pull/17

**Comment by tstellar** [comment]
> Where are you seeing mmap of /dev/shm?
The runtime does mmap of /dev/shm in AqlQueue::AllocRegisteredRingBuffer().  It's unclear if this is where @FalconBsp was seeing a failure, but I am seeing mmap() return -EPERM in this function when /dev/shm is mounted with noexec.

**Comment by tstellar** [comment]
Is there any particular reason this function uses /dev/shm?  My simple test case works using anonymous memory instead.

**Comment by tstellar** [comment]
Just sent out a pull request that may solve this issue: https://github.com/RadeonOpenCompute/ROCR-Runtime/pull/28

**Comment by skeelyamd** [comment]
/dev/shm was used because the queue requires the same physical pages mapped to two virtual address ranges and shm accomplishes that without touching the disk.  This mapping is required to work around differences in command ring read/write pointers and HSA read/write indexes.

Some other file could be used but I don't know of any way to achieve this mapping with anonymous.  We also can't afford delays in updates between the two mappings so whatever file is used shouldn't be on disk anywhere.

**Comment by fxkamd** [comment]
As long as there is enough memory for keeping the data in the page cache, using a file on disk shouldn't cause any delays. Otherwise it's more or less like swap, except it uses a named file instead of anonymous swap space.

You can prevent others from using the same file by unlinking it right after mapping.

**Comment by gstoner** [comment]
Before we make any changes,  there need to be performance experiments, to see what happen in the nominal case, and worst case if there page swap.

**Comment by tstellar** [comment]
I just commented on https://github.com/RadeonOpenCompute/ROCT-Thunk-Interface/issues/8.  To summarize:  I'm hitting this issue when running ROCm inside docker since it mount /dev/shm as noexec.  If performance is a concern, I think the best option would be to fallback to files in /tmp when /dev/shm is mounted noexec or does not exist.

**Comment by skeelyamd** [comment]
As far as I know mmap is the only (reasonable) way to get this specific mapping and it requires a file handle.

For a different reason we looked for a way to double map anonymous memory and the only way we could come up with was by locking it with a TTM BO and mapping that.  This wasn't really usable and the KFD isn't setup to do mmap with mandatory placement.  At that time I did some experiments (for IPC signals, Felix may recall this) with shared mappings of ordinary files and there was a very significant delay in the visibility of updates to different VAs, which mapped the same file offset.  We can't afford any or the CP may observe corrupt packet state.

We could try switching to memfd_create.  This appears to be what we want but does use tmpfs which may have similar restrictions.

Is switching to a different file going to resolve the real issue here?  In general I thought secure systems forbid writeable+executable which unfortunately is required for user mode queues and IOMMUv2.  We have a similar issue with any GPU ISA allocations (also writable and executable though those are allocated from the Thunk/KFD).  It seems it would be better to have separate permissions for host, cp, and sq but hardware is lacking (or software depending on how you feel about the old shadow page tables approach).  Also the Thunk/KFD interfaces don't permit enabling execute after allocation, which means that at least while we are loading ISA there will be executable+writable memory (in addition to the queue memory which has to stay write+exec at all times).
